pkg/kubelet/pod_workers.go

```diff
@@ -874,30 +939,121 @@ func (p *podWorkers) allowStaticPodStart(fullname string, uid types.UID) bool {
 	return true
 }
 
-func (p *podWorkers) managePodLoop(podUpdates <-chan podWork) {
-	var lastSyncTime time.Time
-	var podStarted bool
-	for update := range podUpdates {
-		pod := update.Options.Pod
+// cleanupUnstartedPod is invoked if a pod that has never been started receives a termination
+// signal before it can be started.
+func (p *podWorkers) cleanupUnstartedPod(pod *v1.Pod, status *podSyncStatus) {
+	p.cleanupPodUpdates(pod.UID)
 
-		// Decide whether to start the pod. If the pod was terminated prior to the pod being allowed
-		// to start, we have to clean it up and then exit the pod worker loop.
-		if !podStarted {
-			canStart, canEverStart := p.allowPodStart(pod)
-			if !canEverStart {
-				p.completeUnstartedTerminated(pod)
-				if start := update.Options.StartTime; !start.IsZero() {
-					metrics.PodWorkerDuration.WithLabelValues("terminated").Observe(metrics.SinceInSeconds(start))
-				}
-				klog.V(4).InfoS("Processing pod event done", "pod", klog.KObj(pod), "podUID", pod.UID, "updateType", update.WorkType)
-				return
-			}
-			if !canStart {
-				klog.V(4).InfoS("Pod cannot start yet", "pod", klog.KObj(pod), "podUID", pod.UID)
-				continue
-			}
-			podStarted = true
+	if status.terminatingAt.IsZero() {
+		klog.V(4).InfoS("Pod worker is complete but did not have terminatingAt set, likely programmer error", "pod", klog.KObj(pod), "podUID", pod.UID)
+	}
+	if !status.terminatedAt.IsZero() {
+		klog.V(4).InfoS("Pod worker is complete and had terminatedAt set, likely programmer error", "pod", klog.KObj(pod), "podUID", pod.UID)
+	}
+	status.finished = true
+	status.working = false
+	status.terminatedAt = time.Now()
+
+	if p.startedStaticPodsByFullname[status.fullname] == pod.UID {
+		delete(p.startedStaticPodsByFullname, status.fullname)
+	}
+}
+
+// startPodSync is invoked by each pod worker goroutine when a message arrives on the pod update channel.
+// This method consumes a pending update, initializes a context, decides whether the pod is already started
+// or can be started, and updates the cached pod state so that downstream components can observe what the
+// pod worker goroutine is currently attempting to do. If ok is false, there is no available event. If any
+// of the boolean values is false, ensure the appropriate cleanup happens before returning.
+func (p *podWorkers) startPodSync(podUID types.UID) (ctx context.Context, update podWork, canStart, canEverStart, ok bool) {
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+
+	// verify we are known to the pod worker still
+	status, ok := p.podSyncStatuses[podUID]
```

Comment:
assert in the "fail" sense, or log a programmer error?  i'd love to verify it during unit test, flakiness of working being correct is a key test correctness problem too (because of drain worker)

Reply:
I was referring to logging a programmer error, I think it's closest thing we have to asserts. If we see it somewhere like in the e2e logs, it will be signal something is wrong.

Reply:
I would love to have something that failed/warned a run if a log line was present.  Could also just highlight them using the prow regex.